### PR TITLE
SCANOSS Scanner Improvements

### PR DESCRIPTION
**Pull request by isasmendiagus**
### Corrected Path Mapping

* **Before:** Source location path showed the OSS file while snippet location showed SCANOSS file content URL `https://api.scanoss.com/file_contents/...`

* **After:** Source location now correctly shows the scanned file and snippet location shows the actual matched OSS file
See atached image for check diff.

*See the attached diff image for a visual representation of the path mapping changes*
image

### Fixed Snippet Generation Logic:

* **Before:** Snippets were generated using a cartesian product of PURLs × line ranges, creating duplicate snippets for the same code match.

* **After:** One snippet is created per line range, using the first PURL as primary identifier and storing all PURLs in metadata.

### CODE REVIEW

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 package org.ossreviewtoolkit.plugins.scanners.scanoss
 
+import com.scanoss.dto.LicenseDetails
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:22]: Similar remarks apply to other commit messages in this PR.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 2]
```diff
@@ -135,34 +147,37 @@ private fun getCopyrightFindings(details: ScanFileDetails): List<CopyrightFindin
...
 
-    return buildSet {
-        purls.forEach { purl ->
-            locations.forEach { snippetLocation ->
-                add(Snippet(score, snippetLocation, provenance, purl, license))
-            }
-        }
+    // Store all PURLs in additionalData to preserve the complete information.
+    val additionalData = mapOf(
+        "all_purls" to purls.joinToString(", ")
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:173]: How about instead using the purls as keys, and use empty values for `additionalData`? That way purls can be retrieved back easier than parsing a value string.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 3]
```diff
@@ -135,34 +147,37 @@ private fun getCopyrightFindings(details: ScanFileDetails): List<CopyrightFindin
...
+
+    // Create one snippet per location, using the first PURL as the primary identifier.
+    return ossLocations.map { snippetLocation ->
+        Snippet(score, snippetLocation, provenance, purls.firstOrNull().orEmpty(), license, additionalData)
     }
 }
 
 /**
- * Split [lineRanges] returned by ScanOSS such as "32-105,117-199" into [TextLocation]s for the given [file].
+ * Split [lineRanges] returned by SCANOSS such as "32-105,117-199" into [TextLocation]s for the given [file].
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:183]: As you seem to change this in several places, it indeed pays off to extract this to a dedicated commit.

File: plugins/scanners/scanoss/src/test/assets/scanoss-multiple-purls.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      },
+      "source_hash": "45dd1e50621a8a32f88fbe0251a470ab",
+      "status": "pending",
+      "url": "https://github.com/kdrag0n/proton_bluecross",
+      "url_hash": "a9c1c67f0930dc42dbd40c29e565bcdd",
+      "vendor": "kdrag0n",
+      "version": "15"
+    }
+  ]
+}
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/assets/scanoss-multiple-purls.json:60]: Nit: Missing final newline.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt (hunk 2 above)
Comment (sschuberth) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:173]: Also, the first purl IMO should be omitted here, as that one is already used as the primary purl for the snippet below.

Comment (sschuberth) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:173]: Maybe an approach like

```kotlin
val purls = requireNotNull(details.purls).toMutableList()
...
val primaryPurl = purls.removeFirstOrNull().orEmpty()
```

works, and afterwards put the remaining `purls` into `additionalData`.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 5]
```diff
@@ -118,9 +125,7 @@ private fun getLicenseFindings(details: ScanFileDetails): List<LicenseFinding> {
         LicenseFinding(
             license = validatedLicense,
             location = TextLocation(
-                path = path,
-                startLine = TextLocation.UNKNOWN_LINE,
-                endLine = TextLocation.UNKNOWN_LINE
+                path = path, startLine = TextLocation.UNKNOWN_LINE, endLine = TextLocation.UNKNOWN_LINE
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:128]: Please keep the formatting if otherwise unchanged.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 6]
```diff
@@ -137,9 +142,7 @@ private fun getCopyrightFindings(details: ScanFileDetails): List<CopyrightFindin
         CopyrightFinding(
             statement = copyright.name,
             location = TextLocation(
-                path = path,
-                startLine = TextLocation.UNKNOWN_LINE,
-                endLine = TextLocation.UNKNOWN_LINE
+                path = path, startLine = TextLocation.UNKNOWN_LINE, endLine = TextLocation.UNKNOWN_LINE
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:145]: See above.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 7]
```diff
@@ -217,8 +220,6 @@ fun getUniqueLicenseExpression(licensesDetails: List<LicenseDetails>): SpdxExpre
         return SpdxLicenseIdExpression(SpdxConstants.NOASSERTION)
     }
 
-    return licensesDetails
-        .map { license -> SpdxExpression.parse(license.name) }
-        .reduce { acc, expr -> acc and expr }
+    return licensesDetails.map { license -> SpdxExpression.parse(license.name) }.reduce { acc, expr -> acc and expr }
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:223]: See above.

File: plugins/scanners/scanoss/src/test/assets/scanoss-identified-snippet.json [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+      },
+      "source_hash": "45dd1e50621a8a32f88fbe0251a470ab",
+      "status": "pending",
+      "url": "https://github.com/kdrag0n/proton_bluecross",
+      "url_hash": "a9c1c67f0930dc42dbd40c29e565bcdd",
+      "vendor": "kdrag0n",
+      "version": "15"
+    }
+  ]
+}
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/assets/scanoss-identified-snippet.json:133]: Missing final newline.

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt [hunk 9]
```diff
@@ -208,5 +208,19 @@ class ScanOssResultParserTest : WordSpec({
                     TextLocation("kernel/hung_task.c", 86, 107)
             }
         }
+
+        "should exclude identified snippets from snippet findings" {
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt:212]: "should" should not be part of the test name, as the surrounding container already uses the `should` keyword`.

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt [hunk 10]
```diff
@@ -162,11 +162,51 @@ class ScanOssResultParserTest : WordSpec({
...
+                            mapOf(
+                                "all_purls" to "pkg:github/vdurmont/semver4j"
+                            )
                         )
                     )
                 )
             )
         }
+
+        "should handle multiple PURLs by selecting first as primary and preserving all in metadata" {
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt:175]: "should" should not be part of the test name, as the surrounding container already uses the `should` keyword`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt [hunk 11]
```diff
@@ -39,9 +41,44 @@ import org.ossreviewtoolkit.model.SnippetFinding
 import org.ossreviewtoolkit.model.TextLocation
 import org.ossreviewtoolkit.model.VcsInfo
 import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.utils.spdx.SpdxConstants
 import org.ossreviewtoolkit.utils.spdx.SpdxExpression
+import org.ossreviewtoolkit.utils.spdx.SpdxLicenseIdExpression
 
 class ScanOssResultParserTest : WordSpec({
+    "getUniqueLicenseDetails()" should {
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt:49]: I guess this whole test could be dropped when just using `Collection<SpdxExpression>.toExpression()` instead of introducing this new function.

**Comment by nnobelis** [comment]
@isasmendiagus Just a question since your are on the improvements. Is the anonymous mode still on your radar ?

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt [hunk 12]
```diff
@@ -125,11 +162,65 @@ class ScanOssResultParserTest : WordSpec({
...
                         )
                     )
                 )
             )
         }
+
+        "should handle multiple PURLs by selecting first as primary and preserving all in metadata" {
+            val results = File("src/test/assets/scanoss-multiple-purls.json").readText().let {
+                JsonUtils.toScanFileResultsFromObject(JsonUtils.toJsonObject(it))
+            }
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt:178]: By now, tests assets have been moved to resources in `main`, so you need to rebase and adapt this code as show e.g. [here](https://github.com/oss-review-toolkit/ort/pull/10318/files#diff-abf375b12f262cb8f7df2d3b00b86e7b8d860c2bc25abec91c73f61b7ee243faR47).

**Comment by isasmendiagus** [comment]
> @isasmendiagus Just a question since your are on the improvements. Is the anonymous mode still on your radar ?

Hi @nnobelis, yes. We're about to release the scanoss.java library with path anonymization. Then adding the feature to ORT should be straight forward. 

Question here about priorities, do you prefer the generic report first and then path anonymization? Or the other way around?

**Comment by sschuberth** [comment]
> For us, path anonymization is more important, if @sschuberth can live with delaying the generic report a bit more....

Fine with me.

**Comment by sschuberth** [comment]
@isasmendiagus, please remember to re-request a review from my side if you're done with the updates to address review comment. Otherwise I'm unsure whether the PR is in an intermediate state or not.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 13]
```diff
@@ -144,9 +144,9 @@ private fun getSnippets(details: ScanFileDetails): Set<Snippet> {
     val url = requireNotNull(details.url)
     val purls = requireNotNull(details.purls)
 
-    val licenses = details.licenseDetails.orEmpty().mapTo(mutableSetOf()) { license ->
-        SpdxExpression.parse(license.name)
-    }
+    val license = details.licenseDetails.orEmpty()
+        .map { license -> SpdxExpression.parse(license.name) }
+        .toExpression()?.sorted() ?: SpdxLicenseIdExpression(SpdxConstants.NOASSERTION)
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:170]: This seems to be an optimization that's unrelated to the strict topic of this commit. I'd extract it to a preceding commit saying

```
chore(scanoss): Determine the snippet license only once

The `licenses` variable is constant in the loop, so extract it to determine
the resulting `license` only once.
```

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt [hunk 14]
```diff
@@ -131,5 +134,56 @@ class ScanOssResultParserTest : WordSpec({
...
+            val time = Instant.now()
+            val summary = generateSummary(time, time, results)
+
+            // Verify the snippet finding.
+            summary.snippetFindings should haveSize(1)
+            val snippet = summary.snippetFindings.first().snippets.first()
+
+            // Consolidate the license into a single expression
+            // even though it came from both "scancode" and "file_header" sources.
+            snippet.license shouldBe SpdxExpression.parse("LGPL-2.1-or-later")
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt:193]: You can use the `toSpdx()` helper here like `"LGPL-2.1-or-later".toSpdx()`.

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt [hunk 15]
```diff
@@ -131,5 +134,56 @@ class ScanOssResultParserTest : WordSpec({
...
+            with(summary.snippetFindings.first()) {
+                sourceLocation.path shouldBe "src/check_error.c"
+                sourceLocation.startLine shouldBe 16
+                sourceLocation.endLine shouldBe 24
+            }
+        }
+
+        "handle empty license array with NOASSERTION" {
+            // When a component has an empty licenses array, use NOASSERTION.
+
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssResultParserTest.kt:205]: IMO this does not add much value over the test title, and could be removed.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt (hunk 13 above)
Comment (sschuberth) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:170]: To clarify, I've checked that the added tests also pass without this change. While it's good to have more test coverage, I'm unsure whether that's what you intended to do.

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssScannerDirectoryTest.kt [hunk 16]
```diff
@@ -103,7 +103,7 @@ class ScanOssScannerDirectoryTest : StringSpec({
                         Snippet(
                             99.0f,
                             TextLocation(
-                                "https://osskb.org/api/file_contents/871fb0c5188c2f620d9b997e225b0095",
+                                "utils/src/main/kotlin/random-data-05-06-11.kt", // This is the remote oss filepath
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssScannerDirectoryTest.kt:106]: Consider omitting the comment.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 17]
```diff
@@ -22,6 +22,7 @@ package org.ossreviewtoolkit.plugins.scanners.scanoss
 import com.scanoss.dto.ScanFileDetails
 import com.scanoss.dto.ScanFileResult
 import com.scanoss.dto.enums.MatchType
+import com.scanoss.dto.enums.StatusType
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:25]: Commit message: Duplicate backtick after `generateSummary()`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt (hunk 13 above)
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:170]: > This seems to be an optimization that's unrelated to the strict topic of this commit. I'd extract it to a preceding commit saying

I'll reword it as suggested.

Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:170]: > To clarify, I've checked that the added tests also pass without this change. While it's good to have more test coverage, I'm unsure whether that's what you intended to do.

Tests were included to verify edge cases with duplicate licenses from differences sources and empty license arrays. 

I'll move them to a separate commit preceding the changes with this message:

```
test(scanoss): Add tests for license handling edge cases

Add test cases to verify behavior when combining the same license from 
different sources and handling empty license arrays with NOASSERTION value.
```

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 18]
```diff
@@ -134,32 +146,33 @@ private fun getCopyrightFindings(details: ScanFileDetails): List<CopyrightFindin
...
 
 /**
- * Get the snippet findings from the given [details]. If a snippet returned by ScanOSS contains several Purls,
- * several snippets are created in ORT each containing a single Purl.
+ * Get the snippet findings from the given [details]. If a snippet returned by ScanOSS contains several PURLs,
+ * the function extracts the first PURL as the primary identifier while storing the remaining PURLs in additionalData
+ * to preserve the complete information.
  */
-private fun getSnippets(details: ScanFileDetails): Set<Snippet> {
+private fun getSnippets(details: ScanFileDetails): List<Snippet> {
```
[the remark this answers is not in the document]
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:160]: The change to List is intentional, we need the ordering to correctly pair each snippet with its source location in the zip operation.

[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:160]: Ah, I remember having asked this [here](https://github.com/oss-review-toolkit/ort/pull/10004#discussion_r1998993460) already. But I'm still not sure whether I'm getting the point:

1. (LinkedHash)Sets do maintain insertion order. So in this case using a set is just a nice way of saying "there can be no duplicates".
2. I'm anyway to seeing through how you can guarantee that for the `zip` operation `sourceLocations` and `snippets` are in corresponding order.

[the remark this answers is not in the document]
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:160]: The correspondence between `sourceLocations` and `snippets` in the current implementation is guaranteed because both collections are created in the same order from the same SCANOSS result.

https://github.com/oss-review-toolkit/ort/blob/1683430489b933e06e01dfe052e8416ed64c3dd2/plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt#L65-L67

Initially, I tried to change as little code as possible by switching from Set to List. But we could take a better approach by changing the function signature completely.

What if we associate `sourceLocation` and `snippet` directly inside a funcion? Would you like that approach?

```kotlin
/**
 * Create snippet findings directly from scan file details.
 */
private fun createSnippetFindings(
    details: ScanFileDetails,
    localFilePath: String
): Set<SnippetFinding> {
    
    //....
    
    // Convert both local and OSS line ranges to source locations
    val sourceLocations = convertLines(localFilePath, localLines)  // From "lines" field (e.g., "16-24,30-35")
    val ossLocations = convertLines(ossFile, ossLines)            // From "oss_lines" field (e.g., "34-42,50-55")

    // Log warning if there's a mismatch between local and OSS locations
    if (sourceLocations.size != ossLocations.size) {
        logger.warn {
            "Unexpected mismatch in '$localFilePath': " +
                "${sourceLocations.size} source locations vs ${ossLocations.size} OSS locations. " 
        }
    }

    // Directly pair source locations with their corresponding OSS locations
    return sourceLocations.zip(ossLocations).mapTo(mutableSetOf()) { (sourceLocation, ossLocation) ->
        SnippetFinding(
            sourceLocation,
            setOf(
                Snippet(
                    location = sourceLocation,
                    purl = primaryPurl,
                    //.....
                )
            )
        )
    }
}
```

We convert both to `locations` using the same logic and then explicitly pair them with zip. Since both fields come from the same SCANOSS result and are processed by the same conversion function, the first local location corresponds to the first OSS location, the second to the second, and so on.

Usage would simplify to:
```kotlin
snippetFindings += createSnippetFindings(details, localFilePath)
```

[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:160]: Yes, I like it, as it looks quite clean to me that way.

File: plugins/scanners/scanoss/src/test/resources/scanMulti/mappings/scanoss-multi-response.json [hunk 19]
```diff
@@ -10,7 +10,7 @@
...
...t\",      \"release_date\": \"2021-03-18\",      \"file\": \"scanner/src/main/kotlin/random-data-05-07-04.kt\",      \"url_hash\": \"37faa38a820322fa93bf7a8fa8290bb8\",      \"file_hash\": \"5c8ab9be40df937e46c53509481107cd\",      \"source_hash\": \"5c8ab9be40df937e46c53509481107cd\",      \"file_url\": \"https://osskb.org/api/file_contents/5c8ab9be40df937e46c53509481107cd\",      \"licenses\": [        {          \"name\": \"Apache-2.0\",          \"patent_hints\": \"yes\",          \"copyleft\": \"no\",          \"checklist_url\": \"https://www.osadl.org/fileadmin/checklists/unreflicenses/Apache-2.0.txt\",          \"osadl_updated\": \"2022-03-17 13:38\",          \"source\": \"file_spdx_tag\"        },        {          \"name\": \"Apache-2.0\",          \"patent_hints\": \"yes\",          \"copyleft\": \"no\",          \"checklist_url\": \"https://www.osadl.org/fileadmin/checklists/unreflicenses/Apache-2.0.txt\",          \"osadl_updated\": \"2022-03-17 13:38\",          \"source\": \"scancode\"        }      ],      \"server\": {        \"version\": \"4.4.2\",        \"kb_version\": {          \"monthly\": \"22.02\",          \"daily\": \"22.03.25\"        }      }    }  ]}",
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/resources/scanMulti/mappings/scanoss-multi-response.json:13]: I believe something went wrong here: This file and the file below should not be edited as part of this commit.

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssScannerDirectoryTest.kt [hunk 20]
```diff
@@ -103,7 +103,7 @@ class ScanOssScannerDirectoryTest : StringSpec({
                         Snippet(
                             99.0f,
                             TextLocation(
-                                "https://osskb.org/api/file_contents/871fb0c5188c2f620d9b997e225b0095",
+                                "examples/example.rules.kts", // This is the remote oss filepath
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/test/kotlin/ScanOssScannerDirectoryTest.kt:106]: Not sure if I had mentioned this before, but IMO the comment should be dropped for consistency.

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 21]
```diff
@@ -23,8 +23,11 @@ import com.scanoss.dto.ScanFileDetails
 import com.scanoss.dto.ScanFileResult
 import com.scanoss.dto.enums.MatchType
 
+import java.lang.invoke.MethodHandles
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:27]: Commit message: Is this really a `fix`, or should it be a `feat`?

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt [hunk 22]
```diff
@@ -63,7 +64,13 @@ internal fun generateSummary(startTime: Instant, endTime: Instant, results: List
 
                 MatchType.snippet -> {
                     val localFile = requireNotNull(result.filePath)
-                    snippetFindings += createSnippetFindings(details, localFile)
+                    if (details.status == StatusType.pending) {
+                        snippetFindings += createSnippetFindings(details, localFile)
+                    } else {
+                        logger.warn { "File '$localFile' is identified, not including on snippet findings" }
```
Comment (sschuberth) [inline plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:70]: Nits:
- "on" -> "in"
- Missing dot at end of sentence.

**Review by sschuberth — CHANGES_REQUESTED** [review]
Thanks, just some final minor tweaks required!

File: plugins/scanners/scanoss/src/test/resources/scanMulti/mappings/scanoss-multi-response.json (hunk 19 above)
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/test/resources/scanMulti/mappings/scanoss-multi-response.json:13]: I understand your concern, but the changes to the test files are necessary because:

1. After modifying `ScanOssResultParser.kt` to use `result.filePath` instead of `details.file`, the existing tests would fail since they expected the previous behavior.

2. The changes in the test JSON files (`scanoss-response.json` and `scanoss-multi-response.json`) were specifically made to differentiate between local paths (what we now display to users) and remote paths

If you prefer, I can separate these changes into two distinct commits: one for the functional modification in the main code and another to update the corresponding tests.

File: plugins/scanners/scanoss/src/test/kotlin/ScanOssScannerDirectoryTest.kt (hunk 20 above)
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/test/kotlin/ScanOssScannerDirectoryTest.kt:106]: Yes, you did mention this before. I agree

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt (hunk 21 above)
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:27]: You raise a good point. I agree that 'fix' isn't the most accurate descriptor. 

This is more of a structural improvement to how we handle match data. I propose changing the commit message to:

```
refactor(scanoss): Improve snippet generation with primary PURL approach

Replace cartesian product approach (one snippet per PURL-location pair)
with one snippet per location using primary PURL as identifier and storing
all related PURLs in metadata.

``` 

I'll proceed if you agree

Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt:27]: You raise a good point. I agree that 'fix' isn't the most accurate descriptor.

This is more of a structural improvement to how we handle match data. I propose changing the commit message to:

```
refactor(scanoss): Improve snippet generation with primary PURL approach

Replace cartesian product approach (one snippet per PURL-location pair)
with one snippet per location using primary PURL as identifier and storing
all related PURLs in metadata.
```

I'll proceed if you agree

File: plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt (hunk 22 above)
Comment (isasmendiagus) [reply plugins/scanners/scanoss/src/main/kotlin/ScanOssResultParser.kt]: Changed also from `warn` to `info` log level

File: plugins/scanners/scanoss/src/test/resources/scanMulti/mappings/scanoss-multi-response.json (hunk 19 above)
Comment (sschuberth) [reply plugins/scanners/scanoss/src/test/resources/scanMulti/mappings/scanoss-multi-response.json:13]: > If you prefer, I can separate these changes into two distinct commits: one for the functional modification in the main code and another to update the corresponding tests.

No, commits should be atomic in the sense that cherry picking any of them individually should not break tests. So if code changes requires test (asset) changes, these should indeed be done in the same commit.

My mistake was that I did restore `plugins/scanners/scanoss/src/test/resources` from the previous commit to check if tests would still pass wiothout the changes, but I did run the wrong test, i.e. `ScanOssResultParserTest` which got touched in this commit, not realizing that it's `ScanOssScannerDirectoryTest` / `ScanOssScannerFileTest` that would fail.

What also confused me is that some UUID only get replaced now, although the anonymization feature was removed before. But that's because due to the bug that's fixed in this commit this did not surface earlier. I've verified that now all UUIDs got replaced.

**Review by sschuberth — APPROVED** [review]
